logdna/logdna.py

```diff
@@ -100,27 +106,26 @@ def flush(self):
                         'tags': self.tags if self.tags else None},
                     stream=True,
                     timeout=self.request_timeout)
-                self.buf = []
-                self.buf_byte_length = 0
+                res.raise_for_status()
+
+                # when no RequestException happened
+                del self.buf[:]
+                self.exception_flag = False
                 if self.flusher:
                     self.flusher.cancel()
                     self.flusher = None
-                self.lock.release()
-                # Ensure messages that could've dropped are appended back onto buf
-                self.buf = self.buf + self.secondary
-                self.secondary = []
         except requests.exceptions.RequestException as e:
             if self.flusher:
                 self.flusher.cancel()
                 self.flusher = None
+            self.exception_flag = True
+            if self.verbose in ['true', 'error', 'err', 'e']:
+                self.internalLogger.debug('Error sending logs %s', e)
+        try:
             self.lock.release()
-            if not self.exception_flag:
-                self.exception_flag = True
-                if self.verbose in ['true', 'error', 'err', 'e']:
-                    logger.error('Error in Request to LogDNA: %s', str(e))
-        else:
-            # when no RequestException happened
-            self.exception_flag = False
+        except threading.ThreadError as e:
+            self.internalLogger.debug("Could not unlock: %s", e)
```

Comment:
Does this exception happen frequently? It seems like a weird thing to log, and if we catch it, do we leave the `self.lock` in a locked state while continuing?

Reply:
I just did not find a way to check if the lock is locked at the time when I am releasing it. I am following this documentation here https://docs.python.org/2/library/threading.html#rlock-objects.

Reply:
I will revert the condition to read easier.

Reply:
Theoretically, the block where  I am trying to release the lock should not be executed without locking the lock.